# Homework (Chapter 6) - 201601639 홍승현

- 연습문제 1, 3, 6, 9

## 연습문제 1

`[예제 6-1]`에서 초기 군집 중심을 z1 = x2, z2 = x3, z3 = x7로 다시 설정했을 때, k-평균의 동작을 예제처럼 보이시오. 다중 시작을 사용한다면 둘 중 어느 해를 취할지 판단하시오.

In [1]:
import numpy as np
from sympy import * # 수식 표현을 위해 임포트

In [2]:
x_train = np.array([[[18], [5]], [[20], [9]], [[20], [14]], [[20], [17]], [[5], [15]], [[9], [15]], [[6], [20]]], dtype=float)
k = 3 # 군집의 개수
n = len(x_train) # 훈련집합의 길이
# c_cluster = np.copy(x_train[:3]) # 기존 예제에서의 군집집합
c_cluster = np.copy(x_train[[1, 2, 6]]) # 바뀐 군집집합

A = np.zeros((k, n))

for i in range(n):
    pprint(Eq(Symbol(f"x_{i}"), Matrix(x_train[i]), evaluate=False))

     ⎡18.0⎤
x₀ = ⎢    ⎥
     ⎣5.0 ⎦
     ⎡20.0⎤
x₁ = ⎢    ⎥
     ⎣9.0 ⎦
     ⎡20.0⎤
x₂ = ⎢    ⎥
     ⎣14.0⎦
     ⎡20.0⎤
x₃ = ⎢    ⎥
     ⎣17.0⎦
     ⎡5.0 ⎤
x₄ = ⎢    ⎥
     ⎣15.0⎦
     ⎡9.0 ⎤
x₅ = ⎢    ⎥
     ⎣15.0⎦
     ⎡6.0 ⎤
x₆ = ⎢    ⎥
     ⎣20.0⎦


In [3]:
restore_A = np.copy(A)
while True:
    for i in range(n):
        min_pos = np.argmin(list(map(lambda x: np.linalg.norm(x-x_train[i]), c_cluster)))
        A[min_pos, i] = 1
    pprint(Eq(Symbol("A"), Matrix(A), evaluate=False))
    if (restore_A == A).all():
        break
    for j in range(k):
        c_cluster[j] = x_train[np.where(A[j] == 1)].mean(axis=0)
        pprint(Eq(Symbol(f"z_{j}"), Matrix(c_cluster[j]), evaluate=False))
    
    J = 0
    for i in range(n):
        for j in range(k):
            J += A[j, i] * np.linalg.norm(x_train[i]-c_cluster[j])
    pprint(Eq(Symbol("J"), J))
    restore_A = np.copy(A)
    A = np.zeros((k, n))

    
    

    ⎡1.0  1.0  0.0  0.0  0.0  0.0  0.0⎤
    ⎢                                 ⎥
A = ⎢0.0  0.0  1.0  1.0  0.0  0.0  0.0⎥
    ⎢                                 ⎥
    ⎣0.0  0.0  0.0  0.0  1.0  1.0  1.0⎦
     ⎡19.0⎤
z₀ = ⎢    ⎥
     ⎣7.0 ⎦
     ⎡20.0⎤
z₁ = ⎢    ⎥
     ⎣15.5⎦
     ⎡6.66666666666667⎤
z₂ = ⎢                ⎥
     ⎣16.6666666666667⎦
J = 16.0959466570308
    ⎡1.0  1.0  0.0  0.0  0.0  0.0  0.0⎤
    ⎢                                 ⎥
A = ⎢0.0  0.0  1.0  1.0  0.0  0.0  0.0⎥
    ⎢                                 ⎥
    ⎣0.0  0.0  0.0  0.0  1.0  1.0  1.0⎦


In [4]:
c_cluster = np.copy(x_train[:3]) # 기존 예제에서의 군집집합
A = np.zeros((k, n))
restore_A = np.copy(A)
while True:
    for i in range(n):
        min_pos = np.argmin(list(map(lambda x: np.linalg.norm(x-x_train[i]), c_cluster)))
        A[min_pos, i] = 1
    pprint(Eq(Symbol("A"), Matrix(A), evaluate=False))
    if (restore_A == A).all():
        break
    for j in range(k):
        c_cluster[j] = x_train[np.where(A[j] == 1)].mean(axis=0)
        pprint(Eq(Symbol(f"z_{j}"), Matrix(c_cluster[j]), evaluate=False))
    
    J = 0
    for i in range(n):
        for j in range(k):
            J += A[j, i] * np.linalg.norm(x_train[i]-c_cluster[j])
    pprint(Eq(Symbol("J"), J))
    restore_A = np.copy(A)
    A = np.zeros((k, n))

    ⎡1.0  0.0  0.0  0.0  0.0  0.0  0.0⎤
    ⎢                                 ⎥
A = ⎢0.0  1.0  0.0  0.0  0.0  0.0  0.0⎥
    ⎢                                 ⎥
    ⎣0.0  0.0  1.0  1.0  1.0  1.0  1.0⎦
     ⎡18.0⎤
z₀ = ⎢    ⎥
     ⎣5.0 ⎦
     ⎡20.0⎤
z₁ = ⎢    ⎥
     ⎣9.0 ⎦
     ⎡12.0⎤
z₂ = ⎢    ⎥
     ⎣16.2⎦
J = 33.7722115099058
    ⎡1.0  0.0  0.0  0.0  0.0  0.0  0.0⎤
    ⎢                                 ⎥
A = ⎢0.0  1.0  1.0  1.0  0.0  0.0  0.0⎥
    ⎢                                 ⎥
    ⎣0.0  0.0  0.0  0.0  1.0  1.0  1.0⎦
     ⎡18.0⎤
z₀ = ⎢    ⎥
     ⎣5.0 ⎦
     ⎡      20.0      ⎤
z₁ = ⎢                ⎥
     ⎣13.3333333333333⎦
     ⎡6.66666666666667⎤
z₂ = ⎢                ⎥
     ⎣16.6666666666667⎦
J = 17.2904773686979
    ⎡1.0  0.0  0.0  0.0  0.0  0.0  0.0⎤
    ⎢                                 ⎥
A = ⎢0.0  1.0  1.0  1.0  0.0  0.0  0.0⎥
    ⎢                                 ⎥
    ⎣0.0  0.0  0.0  0.0  1.0  1.0  1.0⎦


### 결론

- 먼저 가까운 군집에 배정한 결과 위와 같은 A의 값이 나왔으며, 새롭게 계산한 군집의 중심은 위의 `z0`, `z1`, `z2`와 같다.
- 최적화 식 J를 적용하면 `16.0959`의 결과가 나오며, 두 번째 루프를 실행하였을 때 행렬의 값은 동일하기 때문에 바뀌지 않는다.
- 기존의 예제의 경우 J는 `17.2904`가 나오기 때문에 바꿔 적용한 군집을 선택할 것이다. 

## 연습문제 3

`k-평균`과 `k-medoids`의 계산 시간을 분석하시오. 어느 것이 몇 배 정도 더 많은 시간을 쓸지 추정할 수 있는가?

### 답

- `k-평균`의 경우 각 샘플을 가지고 평균을 구한 것을 군집의 중심으로 잡기 때문에, 샘플의 개수만큼의 연산을 하므로 `O(n)`의 시간복잡도가 나타내어진다.
- `k-medoids`의 경우 하나의 샘플이 군집에서 자신을 제외한 모든 샘플들의 거리를 계산하여 합을 구하는데, 하나의 샘플에서 `O(n)`이 소요되고, 이를 모두의 샘플에 적용해야 하므로 `O(n^2)`의 시간복잡도가 나타내어진다.

## 연습문제 6

`그림[6-8]`에는 80개의 샘플이 있다. \[0, 0.1\] 구간을 0, \[0.1, 0.2\]를 1, ... , \[0.9, 1\]을 9로 표시하고 히스토그램 방식을 사용한다.

1. 다음 확률을 구하시오
    - 결합확률 P(x1 = 2, x2 = 3), 조건부 확률 P(x1 = 2 | x2 = 3), 주변 확률 P(x2 = 3)

2. 다음 확률을 구하시오.

    - 결합확률 P(x1 = 3, x2 = 5), 조건부 확률 P(x2 = 5 | x1 = 3), 주변 확률 P(x1 = 3)


1. 다음 확률을 구하시오

    - 결합확률 P(x1 = 2, x2 = 3) = 1/80
    - 조건부 확률 P(x1 = 2 | x2 = 3) = 1/80/5/80 = 1/5
    - 주변 확률 P(x2 = 3) = 5/80

2. 다음 확률을 구하시오

    - 결합확률 P(x1 = 3, x2 = 5) =  3/80
    - 조건부 확률 P(x2 = 5 | x1 = 3) = 3/80/12/80 = 3/12 = 1/4
    - 주변 확률 P(x1 = 3) = 12/80

## 연습문제 9

훈련집합이 다음과 같다. PCA를 이용하여 2차원을 1차원으로 변환하는 행렬을 구하시오.

In [18]:
x_train = np.array([[[1], [2]], [[2], [2]], [[3], [2]], [[3], [3]], [[4], [3]], [[4], [4]], [[5]
, [4]], [[6], [4]]], dtype=float)
avg_vector = np.sum(x_train, axis=0)/x_train.shape[0]


In [19]:
avg_vector

array([[3.5],
       [3. ]])

In [20]:
moved_x_vector = x_train - avg_vector

In [48]:
conv_vector = np.sum(list(map(lambda x: x.dot(x.T), moved_x_vector)), axis=0)/moved_x_vector.shape[0]
conv_vector

array([[2.25 , 1.125],
       [1.125, 0.75 ]])

In [49]:
eigen_value, eigen_vector = np.linalg.eig(conv_vector)

In [63]:
print(f"고유값: {eigen_value[0]}\n고유벡터: {eigen_vector[0]}")

고유값: 2.852081728298996
고유벡터: [ 0.8816746  -0.47185793]
